File: solutions/cpp/protein-translation/1/protein_translation.cpp
#include "protein_translation.h"

#include <stdexcept>
#include <unordered_map>
namespace protein_translation {

    const std::unordered_map<std::string, std::string> codon_map = {
        {"AUG", "Methionine"},
        {"UUU", "Phenylalanine"}, {"UUC", "Phenylalanine"},
        {"UUA", "Leucine"},       {"UUG", "Leucine"},
        {"UCU", "Serine"},        {"UCC", "Serine"},
        {"UCA", "Serine"},        {"UCG", "Serine"},
        {"UAU", "Tyrosine"},      {"UAC", "Tyrosine"},
        {"UGU", "Cysteine"},      {"UGC", "Cysteine"},
        {"UGG", "Tryptophan"},
        {"UAA", "STOP"}, {"UAG", "STOP"}, {"UGA", "STOP"}
    };

    std::vector<std::string> proteins(const std::string& rna) {
        std::vector<std::string> result(0);
        if (rna.size() % 3 != 0) {
            throw std::invalid_argument("rna length must be multiple of 3");
        }
        for (size_t i = 0; i < rna.size(); i += 3) {
            auto amino = codon_map.find(rna.substr(i, 3)) ;
            if (amino == codon_map.end()) {
                throw std::invalid_argument("rna strand contains invalid codon");
            }
            if (amino->second == "STOP") {
                break;
            }
            result.push_back(amino->second);
        }
        return result;
    }

}  // namespace protein_translation

File: solutions/python/protein-translation/1/protein_translation.py
from itertools import takewhile

def proteins(strand):
    return [
        amino for amino in takewhile(lambda x: x, (
            _codon_to_amino(strand[i:i+3]) for i in range(0, len(strand), 3)
        ))
    ]

def _codon_to_amino(codon):
    match codon:
        case 'AUG':
            return 'Methionine'
        case 'UUU' | 'UUC':
            return 'Phenylalanine'
        case 'UUA' | 'UUG':
            return 'Leucine'
        case 'UCU' | 'UCC' | 'UCA' | 'UCG':
            return 'Serine'
        case 'UAU' | 'UAC':
            return 'Tyrosine'
        case 'UGU' | 'UGC':
            return 'Cysteine'
        case 'UGG':
            return 'Tryptophan'
        case 'UAA' | 'UAG' | 'UGA':
            return None
        case _:
            raise ValueError(f'Invalid codon: {codon}')